pkg/network/node/egressip.go

```diff
@@ -141,13 +158,70 @@ func getMarkForVNID(vnid, masqueradeBit uint32) string {
 	return fmt.Sprintf("0x%08x", vnid)
 }
 
+func (eip *egressIPWatcher) runIPAssignmentResync(stopCh <-chan struct{}) {
+	var addrChan chan netlink.AddrUpdate
+	addrSubscribe := func() error {
+		addrChan = make(chan netlink.AddrUpdate)
+		err := netlink.AddrSubscribeWithOptions(addrChan, stopCh, netlink.AddrSubscribeOptions{
+			ErrorCallback: func(err error) {
+				klog.Errorf("Failed during AddrSubscribe callback: %v", err)
+			},
+		})
+		// when AddrSubscribeWithOptions fails it does not close the addrChan
+		if err != nil {
+			close(addrChan)
+		}
+		return err
+	}
+
+	syncIPs := func() {
+		eip.tracker.Lock()
+		defer eip.tracker.Unlock()
+
+		eip.SyncIPAssignments()
+	}
+
+	subscribeErr := addrSubscribe()
+	if subscribeErr != nil {
+		klog.Error("Error during netlink subscribe: %v", subscribeErr)
+	}
+
+	for {
+		select {
+		case a, ok := <-addrChan:
+			if !ok {
+				if subscribeErr = addrSubscribe(); subscribeErr != nil {
```

Comment:
you should put the `stopCh` clause before the `addrChan` one, or else it's possible the subscribe thread will process `stopCh` being closed before this thread does, and then you'll try to resubscribe with an already-closed stop channel and get spurious errors, etc, etc

Reply:
good point! done